### ISSUE TITLE
can graph.push and graph.pull run in parallel? 

### ISSUE THREAD

**Issue by allenling** [PyAV-Org/PyAV#1241]
## Overview

we found push and pull operations do not release GIL, we wonder if there is any way to run filters or graphs in parallel?  

my bad, latest pyav does release GIL while calling push and pull, and we are using the latest stable version, which is 9.0.2.

**Comment by WyattBlue** [comment]
To answer the question in the title, I don't see any reason why it wouldn't work, it's all just Cython code. I suppose it might depend on what technical mechanism you're using for parallelism.

For, GPU encoding. It's just as simple as setting the codec to `h264_nvenc` or whatever... Unless the prebuilt PyAV binary doesn't support your gpu encoder, in which case you would have to build PyAV from source while using an ffmpeg that does have that gpu encoder compiled. 

I find GPU encoding is rarely worth it. Quote from my docs:
> If you want to export to a certain codec that is compatible with your GPU, yes, in some cases, it will go noticeably faster, albeit with some slight quality loss
> However, in most other cases, GPU acceleration won't do anything since analyze and creating new media files are mostly CPU bound. Given how relatively complex enabling GPU acceleration is, it is not recommend for most users

If you care about concurrency or just maximum efficiency in general, you might want to write your code in C and just use directly call the ffmpeg C libraries instead of using PyAV. All PyAV really is is a thin layer around the ffmpeg libs.
